### PR TITLE
Add NEMO OpenACC transformation scripts under examples.

### PR DESCRIPTION
**Pull request by arporter**
This PR brings in the `process_nemo.py` and `kernels_trans.py` scripts that Chris and I have been using to transform the NEMO code base. As discussed with @rupertford, it's still a bit rough but the plan is to move functionality out of the script and into the various transformations (where it makes sense). That's the subject of #852.

### CODE REVIEW

**Comment by arporter** [comment]
Ready for a first look.

**Comment by arporter** [comment]
Taking this back as I need to update the script now we get `Call` nodes for calls (rather than CodeBlocks).

**Comment by arporter** [comment]
Happily, that was very easy. Ready for review again now.

File: examples/nemo/eg3/README.md [hunk 1]
```diff
@@ -2,17 +2,21 @@
...
+Kernel and Data directives to NEMO code. Note, the transformations are
 indicative of what could be done - no claim is made as to the performance of
 the resulting code.
 
-Once you have installed PSyclone, either of the scripts may be supplied to
-PSyclone via the -s option:
+A more sophisticated `kernels_trans.py` script is provided in the
+`../scripts` directory. This is designed to work with the NVIDIA (PGI)
+compiler's 'managed memory' option (`-ta=tesla:managed`) and therefore makes
+no attempt to control data movement to/from the GPU. However, it does
```
Comment (rupertford) [inline examples/nemo/eg3/README.md:14]: I'm not sure you should be saying "However, it does add" as if the previous thing is bad. Probably better to to just say "It also adds"

File: examples/nemo/scripts/process_nemo.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    PARSER.add_argument('input_files', metavar='input_file', type=str,
+                        nargs='+',
+                        help="One or more NEMO pre-processed source files")
+    PARSER.add_argument('-o', dest='out_dir',
+                        help="Destination directory for processed source "
+                        "files")
+    PARSER.add_argument('-s', dest='script_file',
+                        help="PSyclone transformation script")
+    PARSER.add_argument('-x', dest='exit_on_error', action='store_true',
+                        help="eXit immediately if PSyclone fails")
```
Comment (rupertford) [inline examples/nemo/scripts/process_nemo.py:96]: As mentioned before, you could replace the X with x

File: examples/nemo/scripts/process_nemo.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python3
```
Comment (rupertford) [inline examples/nemo/scripts/process_nemo.py:1]: As agreed, I'm going to let this and kernels_trans.py on to master without any changes to make it easier to keep them up to date. For the record though, I think this particular script would be useful to re-work into as a generic batch_process script. I've often had to write something to run over all lfric files, for example. I think the only nemo-specific thing here is the list of excluded files which could be stored in a config file, or similar.

File: examples/nemo/scripts/kernels_trans.py [hunk 4]
```diff
@@ -0,0 +1,732 @@
+# -----------------------------------------------------------------------------
```
Comment (rupertford) [inline examples/nemo/scripts/kernels_trans.py:1]: Wow, that's a big script file, although you did warn me :-). As agreed I am not going to suggest changes to this as the point of this PR is to get it onto master so it is easier to maintain. As expected there are lots of good things in it which we should be able to generalise and provide as useful functionality to users over time.

File: examples/nemo/README.md [hunk 5]
```diff
@@ -43,12 +43,49 @@ in the individual example directories for further details.
...
 1. the Tracer advection benchmark routine (tra_adv), as provided by
-Silvia Mocavero of CMCC and
+   Silvia Mocavero of CMCC and
 2. an unmodified NEMO subroutine computing the horizontal component of
-the lateral tracer mixing trend (traldf_iso).
+   the lateral tracer mixing trend (traldf_iso).
+
+## Scripts
+
+Contains two scripts used to process the NEMO code base and add profiling
```
Comment (rupertford) [inline examples/nemo/README.md:55]: and add -> to add ?

File: examples/nemo/README.md [hunk 6]
```diff
@@ -43,12 +43,49 @@ in the individual example directories for further details.
...
+       positional arguments:
+         input_file      One or more NEMO pre-processed source files
+
+       optional arguments:
+         -h, --help      show this help message and exit
+         -o OUT_DIR      Destination directory for processed source files
+         -s SCRIPT_FILE  PSyclone transformation script
+         -x              eXit immediately if PSyclone fails
+
+   In addition to the command-line flags, the script itself has two variables
```
Comment (rupertford) [inline examples/nemo/README.md:78]: has -> contains ?

File: examples/nemo/README.md [hunk 7]
```diff
@@ -43,12 +43,49 @@ in the individual example directories for further details.
...
+   that may be used to control its behaviour:
+
+   - `EXCLUDED_FILES`: list of filenames that will not have OpenACC directives
+     added.
+   - `PROFILE_ALL`: whether or not to add profiling instrumentation to every
+     file, irrespective of whether the file is listed in `EXCLUDED_FILES`.
+   
+2. `kernels_trans.py` is a PSyclone transformation script that adds
+   OpenACC directives and puts fine-grained profiling instrumentation around
```
Comment (rupertford) [inline examples/nemo/README.md:87]: puts -> places ?

File: examples/nemo/README.md [hunk 8]
```diff
@@ -61,7 +98,8 @@ OpenMP parallelisation of traldf_iso over levels.
 ## Example 3
 
 OpenACC parallelisation of tra_adv using the 'data' and 'kernels'
-directives.
+directives. Makes use of the `kernels_trans.py` script from the `scripts`
```
Comment (rupertford) [inline examples/nemo/README.md:101]: Please explain that there is a local script which adds data and kernels then a managed memory script from the scripts directory which only adds kernels and parallel directives.

File: examples/nemo/eg3/Makefile [hunk 9]
```diff
@@ -77,9 +77,15 @@ kernels:
...
 
+# Example of using the more sophisticated script in ../scripts to
+# transform for OpenACC assuming the use of managed memory.
+# TODO #852 - simplify that script and merge it with the functionality of
+# the scripts in this directory.
 managed_mem:
-	${PSYCLONE} -l all -api "nemo" -s ./kernels_managed_mem_trans.py \
+	${PSYCLONE} -l all -api "nemo" -s ../scripts/kernels_trans.py \
    ../code/tra_adv.F90
+	PSYCLONE="${PSYCLONE}" ../scripts/process_nemo.py \
```
Comment (rupertford) [inline examples/nemo/eg3/Makefile:87]: Could you please explain what is going on here. You are creating a variable called PSYCLONE that contains the command to process traldf_iso.F90?

**Review by rupertford — CHANGES_REQUESTED** [review]
Just a few minor comments and please bring up to date with master (as I took so long to review it!).

File: bin/psyclone [hunk 10]
```diff
@@ -1,8 +1,8 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (arporter) [reply bin/psyclone:1]: I think I must have needed that on Glados but I've undone it now.

File: bin/psyclone [hunk 11]
```diff
@@ -1,8 +1,8 @@
...
 # -----------------------------------------------------------------------------
 # BSD 3-Clause License
 #
-# Copyright (c) 2017, Science and Technology Facilities Council
+# Copyright (c) 2017-2020, Science and Technology Facilities Council.
```
[the remark this answers is not in the document]
Comment (arporter) [reply bin/psyclone:5]: Done.

File: examples/nemo/README.md [hunk 12]
```diff
@@ -43,12 +43,49 @@ in the individual example directories for further details.
...
+       Process the specified NEMO source files using PSyclone
+
+       positional arguments:
+         input_file      One or more NEMO pre-processed source files
+
+       optional arguments:
+         -h, --help      show this help message and exit
+         -o OUT_DIR      Destination directory for processed source files
+         -s SCRIPT_FILE  PSyclone transformation script
+         -x              eXit immediately if PSyclone fails
```
[the remark this answers is not in the document]
Comment (arporter) [reply examples/nemo/README.md:76]: Oh yes, that's true, removed.

File: examples/nemo/eg3/README.md (hunk 1 above)
Comment (arporter) [reply examples/nemo/eg3/README.md:14]: I'm not sure I agree that the previous thing is bad but I think "It also adds" is fine so I've gone with it.

File: examples/nemo/scripts/process_nemo.py (hunk 2 above)
Comment (arporter) [reply examples/nemo/scripts/process_nemo.py:96]: Done.

File: examples/nemo/scripts/process_nemo.py (hunk 3 above)
Comment (arporter) [reply examples/nemo/scripts/process_nemo.py:1]: Agreed. I've created #1142.

File: examples/nemo/README.md (hunk 5 above)
Comment (arporter) [reply examples/nemo/README.md:55]: Done.

File: examples/nemo/README.md (hunk 6 above)
Comment (arporter) [reply examples/nemo/README.md:78]: Done.

File: examples/nemo/README.md (hunk 7 above)
Comment (arporter) [reply examples/nemo/README.md:87]: Done.

File: examples/nemo/README.md (hunk 8 above)
Comment (arporter) [reply examples/nemo/README.md:101]: Done.

File: examples/nemo/eg3/Makefile (hunk 9 above)
Comment (arporter) [reply examples/nemo/eg3/Makefile:87]: That took me a while to work out! I've added a comment and updated the README as well as I'd forgotten to document a feature of the script.

**Comment by arporter** [comment]
Assuming the CI is happy, this is ready for another look from @rupertford.

**Review by rupertford — APPROVED** [review]
All issues have been addressed.

**Pull request merged**